src/typechecker.erl

```diff
@@ -748,6 +748,53 @@ negate_bool_type({atom, P, false}) ->
 negate_bool_type(Ty) ->
     Ty.
 
+-spec is_list_type(type()) -> boolean().
+is_list_type({type, _, Name, _}) ->
+    lists:member(Name, [nil, list, nonempty_list, maybe_improper_list, nonempty_improper_list]);
+is_list_type(_) -> false.
+
+-type list_view() :: {empty | nonempty | any | none, type(), type()}.
+
+-spec list_view(type()) -> false | list_view().
+list_view(Ty = {type, _, T, Args}) when ?is_list_type(Ty) ->
```

Comment:
Will this one replace `expect_list_type/2`?

Reply:
This would make sense, but let's leave it for another PR.